workflow/docker_util.py

```diff
@@ -0,0 +1,201 @@
+import docker
+import sys, re, os, shutil, tempfile, hashlib, json, base64, functools
+from pathlib import PurePath, Path
+
+from py2cytoscape import data
+from workflow import WorkflowRun
+from typing import Dict, List, Optional, Any, Type
+
+from workflow.runner import *
+
+
+CLIENTS = {
+    'TPS'       : 'ajshedivy/tps',
+    'CYTOSCAPE' : 'ajshedivy/tps-cytoscape'
+}
+
+
+DEFAULT_HASH_LENGTH = 7
+DOCKER_OUTFOLDER    = "output/output-{}"
+
+def prepare_path_docker(orig_path: PurePath) -> str:
+    """
+    Prepare an absolute path for mounting as a Docker volume.
+    Converts Windows file separators to posix separators.
+    Converts Windows drive letters in absolute paths.
+    """
+    # TODO consider testing PurePath.is_absolute()
+    prepared_path = orig_path.as_posix()
+    # Check whether the path matches an absolute Windows path with a drive letter
+    match = re.match(r'^([A-Z]:)(.*)', prepared_path)
+    if match:
+        # The first group is the drive such as C:
+        drive = match.group(1).lower()[0]
+        # The second group is the rest of the path such as /Users/me
+        prepared_path = match.group(2)
+        prepared_path = '//' + drive + prepared_path
+    return prepared_path
+
+
+def hash_params_sha1_base32(params_dict: Dict[str, Any], length: Optional[int] = None) -> str:

```

Comment:
I don't think we should create output directory names.  The user should provide the directory name, and we use it as it was provided.

Reply:
I finally remembered why I had to create an output directory relative to tps root directory...

We need to link a volume to the container so that the outputs can be linked to the host file system. If the user does not specify an output folder, then the tps root dir is the default location for all of the workflow output files. This would mean that the root directory would be mounted to the container during the tps step of the workflow. Although this would still work in theory, It may not be a good practice to mount any arbitrary host directory to the container. To account for this, I added the unique output directory with a runtime hash that will be automatically generated within the tps root directory. Only this directory will be mounted to the container (which will be empty). 

There may be better, more secure ways of sharing files between a container and host that we can look into as well.

Reply:
An intermediate solution could be to set a default output directory called `output` without the hash id.  That would still give us a volume to mount and keep things simple.